### PR TITLE
[ML] show expected model outputs for each nlp task type

### PR DESCRIPTION
**Pull request by benwtrent**
### Preview

https://stack-docs_2112.docs-preview.app.elstc.co/guide/en/machine-learning/master/ml-nlp-model-ref.html

### CODE REVIEW

File: docs/en/stack/ml/nlp/ml-nlp-model-ref.asciidoc [hunk 1]
```diff
@@ -36,6 +36,34 @@ refer to <<ml-nlp-overview>>.
...
 * https://huggingface.co/microsoft/mpnet-base
 
+[discrete]
+=== Fill mask expected model output
+
+Fill mask is a specific kind of token classification. Usually the base training task of many transformer models.
+
+For the Elastic stack's fill mask NLP task to understand the model output, it must have a specific format. 
+
+It needs to be <number of sequences inferred> by <number of tokens in sequence> by <token score for all tokens in vocabulary>.
```
Comment (lcawl) [inline docs/en/stack/ml/nlp/ml-nlp-model-ref.asciidoc:46]: > It needs to be <number of sequences inferred> by <number of tokens in sequence> by <token score for all tokens in vocabulary>.

I think this sentence should be joined to the previous paragraph. I also think there's some words/concepts missing here. For example: "It is a nested set of {arrays? objects? tuples?} for each inferred sequence, where every token in every sequence has token scores for each token in the vocabulary.

File: docs/en/stack/ml/nlp/ml-nlp-model-ref.asciidoc [hunk 2]
```diff
@@ -45,6 +73,36 @@ refer to <<ml-nlp-overview>>.
...
+The `"o"` entity label indicates that the current token is outside any entity.
+`"b"` indicates the beginning of an entity.
+`"misc"` is a miscellaneous entity.
+`"loc"` is a location.
+`"per"` is a person.
+`"org"` is an organization.
+
+The response format of the model must be <number of sequences inferred> by <number of tokens in sequence> by <classification score for each entity for that token>.
+
+An example of this is where the order of the entity classification labels is: .
```
Comment (lcawl) [inline docs/en/stack/ml/nlp/ml-nlp-model-ref.asciidoc:92]: I think the intro from the previous example was a better lead-in:
```suggestion
Here is an example with a single sequence `"Waldo is in Paris"`:
```

File: docs/en/stack/ml/nlp/ml-nlp-model-ref.asciidoc [hunk 3]
```diff
@@ -86,6 +154,26 @@ Using `DPREncoderWrapper`:
...
 * https://huggingface.co/cardiffnlp/twitter-roberta-base-sentiment
 
+[discrete]
+=== Text classification expected model output
+
+With text classification, the entire sequence is classified. An example of this is sentiment analysis.
+
+The output of the model must be <number of sequences> <classification score of each sequence>.
+
+Example for a binary classification model of happy, sad:
```
Comment (lcawl) [inline docs/en/stack/ml/nlp/ml-nlp-model-ref.asciidoc:164]:
```suggestion
Here is an example with two sequences for a binary classification model of "happy" and "sad":
```

File: docs/en/stack/ml/nlp/ml-nlp-model-ref.asciidoc [hunk 4]
```diff
@@ -97,3 +185,27 @@ Using `DPREncoderWrapper`:
...
+[discrete]
+=== Zero-shot text classification expected model output
+
+Zero-shot text classification allows text to be classified for arbitrary labels not necessarily part of the original
+training. Each sequence is combined with the label given some hypothesis template. The model then scores each of these
+combinations according to `[entailment, neutral, contradiction]`
+
+The required output of the model is <number of sequences> <number of labels> <classification for label>.
+
+Classifying a single sequence against 4 labels would look like the following:
```
Comment (lcawl) [inline docs/en/stack/ml/nlp/ml-nlp-model-ref.asciidoc:198]:
```suggestion
Here is an example with a single sequence classified against 4 labels:
```

File: docs/en/stack/ml/nlp/ml-nlp-model-ref.asciidoc [hunk 5]
```diff
@@ -45,6 +73,36 @@ refer to <<ml-nlp-overview>>.
...
+ only the following classification labels are supported: "o", "b_misc", "i_misc", "b_per", "i_per", "b_org", "i_org", "b_loc", "i_loc".
+
+The `"o"` entity label indicates that the current token is outside any entity.
+`"b"` indicates the beginning of an entity.
+`"misc"` is a miscellaneous entity.
+`"loc"` is a location.
+`"per"` is a person.
+`"org"` is an organization.
+
+The response format of the model must be <number of sequences inferred> by <number of tokens in sequence> by <classification score for each entity for that token>.
```
Comment (lcawl) [inline docs/en/stack/ml/nlp/ml-nlp-model-ref.asciidoc:90]: > The response format of the model must be <number of sequences inferred> by <number of tokens in sequence> by <classification score for each entity for that token>.

Per my suggestion for the previous section, I suggest using something like this instead: "It is a nested set of {arrays? objects? tuples?} for each inferred sequence, where every token in every sequence has classification scores for each entity label."

File: docs/en/stack/ml/nlp/ml-nlp-model-ref.asciidoc [hunk 6]
```diff
@@ -75,6 +133,16 @@ Using `DPREncoderWrapper`:
 * https://huggingface.co/castorini/bpr-nq-ctx-encoder
 * https://huggingface.co/castorini/bpr-nq-question-encoder
 
+[discrete]
+=== Text embedding expected model output
+
+Text embedding allows for semantic embedding of text for dense information retrieval. 
+
+The required format here is that the model outputs the specific embedding directly without any additional pooling.
```
Comment (lcawl) [inline docs/en/stack/ml/nlp/ml-nlp-model-ref.asciidoc:141]: To align with the wording in the other sections:
```suggestion
The output of the model must be the specific embedding directly without any additional pooling. 
```

File: docs/en/stack/ml/nlp/ml-nlp-model-ref.asciidoc [hunk 7]
```diff
@@ -86,6 +154,26 @@ Using `DPREncoderWrapper`:
 * https://huggingface.co/nateraw/bert-base-uncased-emotion
 * https://huggingface.co/cardiffnlp/twitter-roberta-base-sentiment
 
+[discrete]
+=== Text classification expected model output
+
+With text classification, the entire sequence is classified. An example of this is sentiment analysis.
```
Comment (lcawl) [inline docs/en/stack/ml/nlp/ml-nlp-model-ref.asciidoc:160]: IMO the example here doesn't add a lot of value, but if we keep it, I recommend merging the sentences like this:
```suggestion
With text classification (for example, in tasks like sentiment analysis), the entire sequence is classified.
```

File: docs/en/stack/ml/nlp/ml-nlp-model-ref.asciidoc [hunk 8]
```diff
@@ -86,6 +154,26 @@ Using `DPREncoderWrapper`:
...
+
+The output of the model must be <number of sequences> <classification score of each sequence>.
```
Comment (lcawl) [inline docs/en/stack/ml/nlp/ml-nlp-model-ref.asciidoc:162]: > The output of the model must be <number of sequences> <classification score of each sequence>.

Per my suggestion for the previous sections, I suggest using something like this instead: "It is a nested set of {arrays? objects? tuples?} for each inferred sequence, where every sequence has classification scores for each entity label."

File: docs/en/stack/ml/nlp/ml-nlp-model-ref.asciidoc [hunk 9]
```diff
@@ -97,3 +185,27 @@ Using `DPREncoderWrapper`:
...
 * https://huggingface.co/cross-encoder/nli-roberta-base
+
+[discrete]
+=== Zero-shot text classification expected model output
+
+Zero-shot text classification allows text to be classified for arbitrary labels not necessarily part of the original
+training. Each sequence is combined with the label given some hypothesis template. The model then scores each of these
+combinations according to `[entailment, neutral, contradiction]`
+
+The required output of the model is <number of sequences> <number of labels> <classification for label>.
```
Comment (lcawl) [inline docs/en/stack/ml/nlp/ml-nlp-model-ref.asciidoc:196]: > The required output of the model is <number of sequences> <number of labels> <classification for label>.

Per my suggestion for the previous sections, I suggest using something like this instead: "The required output of the model is a nested set of {arrays? objects? tuples?} for each inferred sequence, where every sequence has "entailment", "neutral", and "contradiction" classification scores for each (entity?) label."

**Review by lcawl — COMMENTED** [review]
I've added some more suggestions and comments...

File: docs/en/stack/ml/nlp/ml-nlp-model-ref.asciidoc (hunk 1 above)
Comment (benwtrent) [reply docs/en/stack/ml/nlp/ml-nlp-model-ref.asciidoc:46]: The reader shouldn't be reading this unless they understand machine learning model outputs and are wondering how to provide their own to conform to our pipeline. 

I am not sure we need to be super vocab friendly here.

Technically, it's nested sets of float tensors of specific dimension related to the input and configuration. I will see what other frameworks do for their output pipeline specification documentation.

File: docs/en/stack/ml/nlp/ml-nlp-model-ref.asciidoc [hunk 10]
```diff
@@ -36,6 +36,35 @@ refer to <<ml-nlp-overview>>.
...
+     [ 0, 0, 0, 0, 0, 0, 0 ], // capital
+     [ 0, 0, 0, 0, 0, 0, 0 ], // of
+     [ 0.01, 0.01, 0.3, 0.01, 0.2, 1.2, 0.1 ], // [MASK]
+     [ 0, 0, 0, 0, 0, 0, 0 ], // is
+     [ 0, 0, 0, 0, 0, 0, 0 ] // Paris
+   ] 
+]
+----
+
+The predicted value here for `[MASK]` is obviously `"France"` with a score of 1.2.
```
Comment (lcawl) [inline docs/en/stack/ml/nlp/ml-nlp-model-ref.asciidoc:66]:
```suggestion
The predicted value here for `[MASK]` is `"France"` with a score of 1.2.
```

File: docs/en/stack/ml/nlp/ml-nlp-model-ref.asciidoc [hunk 11]
```diff
@@ -45,6 +74,37 @@ refer to <<ml-nlp-overview>>.
...
+
+The `"O"` entity label indicates that the current token is outside any entity.
+`"I"` indicates that the token is inside an entity.
+`"B"` indicates the beginning of an entity.
+`"MISC"` is a miscellaneous entity.
+`"LOC"` is a location.
+`"PER"` is a person.
+`"ORG"` is an organization.
+
+The response format must be a float tensor with `shape(<number of sequences>, <number of tokens>, <num classification labels>)`.
```
Comment (lcawl) [inline docs/en/stack/ml/nlp/ml-nlp-model-ref.asciidoc:92]: To match the other two, I think "num" should be "number":
```suggestion
The response format must be a float tensor with `shape(<number of sequences>, <number of tokens>, <number of classification labels>)`.
```

File: docs/en/stack/ml/nlp/ml-nlp-model-ref.asciidoc [hunk 12]
```diff
@@ -86,6 +155,25 @@ Using `DPREncoderWrapper`:
 * https://huggingface.co/nateraw/bert-base-uncased-emotion
 * https://huggingface.co/cardiffnlp/twitter-roberta-base-sentiment
 
+[discrete]
+=== Text classification expected model output
+
+With text classification (for example, in tasks like sentiment analysis), the entire sequence is classified. The output of
+the model must be a float tensor with `shape(<number of sequences>, <num classification labels>)`.
```
Comment (lcawl) [inline docs/en/stack/ml/nlp/ml-nlp-model-ref.asciidoc:162]:
```suggestion
the model must be a float tensor with `shape(<number of sequences>, <number of classification labels>)`.
```

File: docs/en/stack/ml/nlp/ml-nlp-model-ref.asciidoc [hunk 13]
```diff
@@ -97,3 +185,26 @@ Using `DPREncoderWrapper`:
...
 * https://huggingface.co/cross-encoder/nli-distilroberta-base
 * https://huggingface.co/cross-encoder/nli-roberta-base
+
+[discrete]
+=== Zero-shot text classification expected model output
+
+Zero-shot text classification allows text to be classified for arbitrary labels not necessarily part of the original
+training. Each sequence is combined with the label given some hypothesis template. The model then scores each of these
+combinations according to `[entailment, neutral, contradiction]`. The output of the model must be a float tensor 
+with `shape(<number of sequences>, <number of labels>, <3>)`.
```
Comment (lcawl) [inline docs/en/stack/ml/nlp/ml-nlp-model-ref.asciidoc:195]: If the angular brackets are intended to denote variable values and 3 is not a variable, I think we should remove the brackets:
```suggestion
with `shape(<number of sequences>, <number of labels>, 3)`.
```

**Review by lcawl — APPROVED** [review]
Three minor suggestions, otherwise LGTM

File: docs/en/stack/ml/nlp/ml-nlp-model-ref.asciidoc (hunk 10 above)
Comment (benwtrent) [reply docs/en/stack/ml/nlp/ml-nlp-model-ref.asciidoc:66]: `obviously` too snarky? :D

Comment (lcawl) [reply docs/en/stack/ml/nlp/ml-nlp-model-ref.asciidoc]: Yeah, don't want people to feel bad if like me they take a moment to see which vocabulary item has the highest score

**Review by davidkyle — COMMENTED** [review]
My preference would be not to have the output description interleaved in the list of models. Could they go in a separate section at the end so the list is supported models is uninterrupted and easier to scan. That list of models is the main reason people visit this page

**Comment by benwtrent** [comment]
@lcawl what do you think? ^

**Review by davidkyle — APPROVED** [review]
LGTM

**Pull request merged**